File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int total = 0, sum = 0;
        unordered_map<int, int> sumFreq;
        sumFreq[0] = 1;
        for(int i=0; i<nums.size(); i++){
            sum += nums[i];
            if(sumFreq.find(sum - k) != sumFreq.end()){
                total += sumFreq[sum - k];
            }
            sumFreq[sum]++;
        }
        return total;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int total = 0;
        HashMap<Integer, Integer> sumArray = new HashMap<>();
        int sum = 0;
        sumArray.put(0, 1);
        for(int i=0; i<nums.length; i++){
            sum += nums[i];
            if(sumArray.containsKey(sum - k)){
                total += sumArray.get(sum-k);
            }
            sumArray.put(sum, sumArray.getOrDefault(sum, 0)+1);
        }
        return total;
    }
}